vulkan/vulkan_device.cc

```diff
@@ -0,0 +1,320 @@
+}
+
+const VulkanHandle<VkDevice>& VulkanDevice::Handle() const {
+  return device_;
+}
+
+const VulkanHandle<VkPhysicalDevice>& VulkanDevice::PhysicalDeviceHandle()
+    const {
+  return physical_device_;
+}
+
+const VulkanHandle<VkQueue>& VulkanDevice::QueueHandle() const {
+  return queue_;
+}
+
+const VulkanHandle<VkCommandPool>& VulkanDevice::CommandPool() const {
+  return command_pool_;
+}
+
+uint32_t VulkanDevice::GraphicsQueueIndex() const {
+  return graphics_queue_index_;
+}
+
+bool VulkanDevice::GetSurfaceCapabilities(
+    const VulkanSurface& surface,
+    VkSurfaceCapabilitiesKHR* capabilities) const {
+  if (!surface.IsValid()) {
+    return false;
+  }
+
+  return VK_CALL_LOG_ERROR(vk.GetPhysicalDeviceSurfaceCapabilitiesKHR(
+             physical_device_, surface.Handle(), capabilities)) == VK_SUCCESS;
+}
+
+bool VulkanDevice::GetPhysicalDeviceFeatures(
+    VkPhysicalDeviceFeatures* features) const {
+  if (features == nullptr || !physical_device_) {
+    return false;
+  }
+  vk.GetPhysicalDeviceFeatures(physical_device_, features);
+  return true;
+}
+
+bool VulkanDevice::GetPhysicalDeviceFeaturesSkia(uint32_t* sk_features) const {
+  if (sk_features == nullptr) {
+    return false;
+  }
+
+  VkPhysicalDeviceFeatures features = {0};
+  if (!GetPhysicalDeviceFeatures(&features)) {
+    return false;
+  }
+
+  uint32_t flags = 0;
+
+  if (features.geometryShader) {
+    flags |= kGeometryShader_GrVkFeatureFlag;
+  }
+  if (features.dualSrcBlend) {
+    flags |= kDualSrcBlend_GrVkFeatureFlag;
+  }
+  if (features.sampleRateShading) {
+    flags |= kSampleRateShading_GrVkFeatureFlag;
+  }
+
+  *sk_features = flags;
+  return true;
+}
+
+std::vector<VkQueueFamilyProperties> VulkanDevice::GetQueueFamilyProperties()
+    const {
+  uint32_t count = 0;
+
+  vk.GetPhysicalDeviceQueueFamilyProperties(physical_device_, &count, nullptr);
+
+  std::vector<VkQueueFamilyProperties> properties;
+  properties.resize(count, {0});
+
+  vk.GetPhysicalDeviceQueueFamilyProperties(physical_device_, &count,
+                                            properties.data());
+
+  return properties;
+}
+
+bool VulkanDevice::ChooseSurfaceFormat(const VulkanSurface& surface,
+                                       VkSurfaceFormatKHR* format) const {
+  if (!surface.IsValid() || format == nullptr) {
+    return false;
+  }
+
+  uint32_t format_count = 0;
+  if (VK_CALL_LOG_ERROR(vk.GetPhysicalDeviceSurfaceFormatsKHR(
+          physical_device_, surface.Handle(), &format_count, nullptr)) !=
+      VK_SUCCESS) {
+    return false;
+  }
+
+  if (format_count == 0) {
+    return false;
+  }
+
+  VkSurfaceFormatKHR formats[format_count];
+  if (VK_CALL_LOG_ERROR(vk.GetPhysicalDeviceSurfaceFormatsKHR(
+          physical_device_, surface.Handle(), &format_count, formats)) !=
+      VK_SUCCESS) {
+    return false;
+  }
+
+  for (uint32_t i = 0; i < format_count; i++) {
+    if (formats[i].format == VK_FORMAT_R8G8B8A8_UNORM) {
```

Comment:
I wasn't sure about Skias expectations either so I just picked the value corresponding to `kRGBA_8888_GrPixelConfig` in `GrVkFormatToPixelConfig` in the Skia sources. So I am assuming Skia want explicit control over the same. I will conform and rewrite this selection logic to make it more generic in a subsequent patch.